File: java version(old)/important/BinaryTreeLevelOrderTraversalII.java
// Tree, Breadth-First Search, Binary Tree
// Amazon 4 Apple 4 Microsoft 3
// https://leetcode.com/problems/binary-tree-level-order-traversal-ii/

// Given the root of a binary tree, return the bottom-up level order traversal of its nodes' values. (i.e., from left to right, level by level from leaf to root).

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class BinaryTreeLevelOrderTraversalII {

  public List<List<Integer>> levelOrderBottom(TreeNode root) {
    LinkedList<List<Integer>> result = new LinkedList<>();
    if (root == null) {
      return result;
    }

    Queue<TreeNode> queue = new LinkedList<>();
    queue.offer(root);
    while (!queue.isEmpty()) {
      List<TreeNode> currentLevelNodes = this.getCurrentLevelNodes(queue);
      List<Integer> currentLevelResult = new ArrayList<>();
      for (TreeNode node : currentLevelNodes) {
        currentLevelResult.add(node.val);
        if (node.left != null) {
          queue.offer(node.left);
        }
        if (node.right != null) {
          queue.offer(node.right);
        }
      }
      result.addFirst(currentLevelResult);
    }

    return result;
  }

  private List<TreeNode> getCurrentLevelNodes(Queue<TreeNode> queue) {
    List<TreeNode> nodes = new ArrayList<>();
    while (!queue.isEmpty()) {
      nodes.add(queue.poll());
    }

    return nodes;
  }
}

File: P1/BinaryTreeLevelOrderTraversalII.py
# Tree, Breadth-First Search, Binary Tree
# Amazon 4 Apple 4 Microsoft 3
# https://leetcode.com/problems/binary-tree-level-order-traversal-ii/

# Given the root of a binary tree, return the bottom-up level order traversal of its nodes' values. (i.e., from left to right, level by level from leaf to root).

from collections import deque
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class BinaryTreeLevelOrderTraversalII:
    def levelOrderBottom(self, root: Optional[TreeNode]) -> List[List[int]]:
        result = deque()
        if not root:
            return result

        queue = deque()
        queue.append(root)
        while queue:
            currentLevelNodes = self.getCurrentLevelNodes(queue)
            currentLevelResult = list()
            for node in currentLevelNodes:
                currentLevelResult.append(node.val)
                if node.left:
                    queue.append(node.left)

                if node.right:
                    queue.append(node.right)

            result.appendleft(currentLevelResult)

        return result

    def getCurrentLevelNodes(self, queue) -> List[TreeNode]:
        nodes = list()
        while queue:
            nodes.append(queue.popleft())

        return nodes
